src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,491 @@
+                }
+
+                projectDiagnosticTasks.Add(new KeyValuePair<ProjectId, Task<AnalysisResult>>(
+                    project.Id,
+                    GetProjectAnalysisResultAsync(
+                        analyzers,
+                        project,
+                        options,
+                        cancellationToken)));
+            }
+
+            ImmutableDictionary<ProjectId, AnalysisResult>.Builder projectDiagnosticBuilder = ImmutableDictionary.CreateBuilder<ProjectId, AnalysisResult>();
+            foreach (var task in projectDiagnosticTasks)
+            {
+                var result = await task.Value.ConfigureAwait(false);
+                if (result == null)
+                {
+                    continue;
+                }
+
+                projectDiagnosticBuilder.Add(task.Key, result);
+            }
+
+            return projectDiagnosticBuilder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Returns a list of all analysis results inside the specific project. This is an asynchronous operation.
+        /// </summary>
+        /// <param name="analyzers">The list of analyzers that should be used</param>
+        /// <param name="project">The project that should be analyzed
+        /// <see langword="false"/> to use the behavior configured for the specified <paramref name="project"/>.</param>
+        /// <param name="cancellationToken">The cancellation token that the task will observe.</param>
+        /// <returns>A list of analysis results inside the project</returns>
+        private static async Task<AnalysisResult> GetProjectAnalysisResultAsync(
+            ImmutableArray<DiagnosticAnalyzer> analyzers,
+            Project project,
+            Options analyzerOptionsInternal,
+            CancellationToken cancellationToken)
+        {
+            WriteLine($"Running analyzers for {project.Name}", ConsoleColor.Gray);
+
+            // update the project compilation options
+            var modifiedSpecificDiagnosticOptions = project.CompilationOptions.SpecificDiagnosticOptions.Add("AD0001", ReportDiagnostic.Error);
+            // Report exceptions during the analysis process as errors
+            var modifiedCompilationOptions = project.CompilationOptions.WithSpecificDiagnosticOptions(modifiedSpecificDiagnosticOptions);
+            var processedProject = project.WithCompilationOptions(modifiedCompilationOptions);
+
+            try
+            {
+                Compilation compilation = await processedProject.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var newCompilation = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(compilation.SyntaxTrees);
+
+                // TODO some analyzer require additionaltext, we need to think about how to support such analyzer.
+                // https://github.com/dotnet/roslyn/issues/23108
+                var analyzerOptions = new AnalyzerOptions(ImmutableArray.Create<AdditionalText>());
```

Comment:
We should use [Project.AnalyzerOptions](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/Project.cs,6d0873d680bca897,references), which already should include additional text files.